### PR TITLE
Inspect dataset structure and propose way to use it for our evaluation #14

### LINKED ISSUE #14: Inspect dataset structure and propose way to use it for our evaluation
### 🎟️ Inspect Dataset & Define Structure

**Description**  
Once the dataset has been downloaded ([see [ticket](https://github.com/KamaTechOrg/AgStream/issues/8)]), the next step is to **inspect its structure** and decide how we will use it for training and evaluation.  
Key questions to resolve:  
- Do we need to evaluate on the entire dataset, or only a split?  
- Are there predefined **train/val/test** splits, or do we need to create them?  
- How can we map each image to its correct **label**?  
- Should we convert the dataset into a simple **classification-style directory structure** such as:  
  ```
  images/
  ├── 0/
  │   ├── image_1.png
  │   ├── image_2.png
  ├── 1/
  │   ├── image_1.png
  │   ├── image_2.png
  ```

---

#### DoD
- [ ] Explore dataset contents and document whether predefined splits exist  
- [ ] Identify how labels are provided (e.g., folder names, CSV file, annotation file)  
- [ ] Decide if we keep the dataset structure as-is or convert it to a classification-style folder format  
- [ ] Update `docs/dataset_preparation.md` with:  
  - Explanation of the dataset’s native format  
  - Label mapping description  
  - Decision on split strategy (use existing / generate custom)  
- [ ] ~If restructuring is needed, update `scripts/preprocess_data.sh` to enforce the new format~  

---

#### 🎯 Acceptance Criteria
- [ ] `docs/dataset_prep.md` clearly explains dataset structure, splits, and label mapping  
- [ ] Final decision documented on whether to restructure into classification-style directories  
- [ ] If required, preparation script produces correct folder layout for training and validation  

---

**Labels:** `dataset`, `structure`, `preparation`  
**Reviewers:** Another student pair

### PR DESCRIPTION
**Pull request by Rut-Vahab**
# Dataset Inspection & Preparation

## 1. Original Structure
- Files are provided in **segmentation** format: images + masks.  
- **Labels** are given as `LabelID` in a CSV file.  
- **No** predefined split into train/val/test exists.  

## 2. Label Mapping
- A mapping was created between `LabelID` and class names (from `datasets.py`).  
- Found **85 active classes** out of 101.  
- Removed unused classes to simplify the workflow.  

## 3. Conversion to Classification Structure
- A new structure was created in **classification-style** format:  
dataset/classifier_data/
├── 0/
│ ├── image_1.png
│ ├── image_2.png
├── 1/
│ ├── image_1.png
│ ├── image_2.png

- Each class has its own folder with its examples.  

## 4. Dataset Splits
- Since no predefined splits exist → we created a custom **train/val** split.  
- The split is applied during preparation without modifying the original dataset.  

## 5. Preparation Script
Added steps to `preprocess_data.sh`:  
1. Download and prepare the original dataset.  
2. Convert to class-based (classification) structure.  

---

## ✅ Acceptance Criteria
- Full documentation of dataset structure.  
- Unused classes removed (85 remained).  
- Dataset converted to classification-friendly structure.  
- Train/val split added.  
- Preparation script (`preprocess_data.sh`) generates the complete format automatically.  
- Decided to use **Macro-F1, Balanced Accuracy, Multiclass MCC** to handle class imbalance.  

---

**Dataset is ready for use in the evaluation pipeline.**

### CODE REVIEW

File: scripts/preprocess_data.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -e
+
+DATA_ROOT="./dataset"
+
+echo "=== Step 1: Downloading/Preparing dataset ==="
+python dataset_preparation/dataset_prep.py --data_root $DATA_ROOT
+
+echo "=== Step 2: Mapping dataset to classification format ==="
+python dataset_preparation/map_dataset.py --dataset_target CropsOrWeed9
```
Comment (r83575) [inline scripts/preprocess_data.sh]: First of all, you don’t need to map it to the CropsOrWeed9 dataset; we use the CropsAndWeed dataset, and the mapping process is not required because it’s already mapped.
In addition, currently it’s problematic to convert it into a classification format, because each image may include multiple types of plants.

File: docs/dataset_preparation.md [hunk 2]
```diff
@@ -1,45 +1,175 @@
...
+### Trade-offs
+- Lost pixel-level information (no semantic masks)  
+- Reduced from **100 → 83 classes** (only populated ones kept)  
+- Single-label classification (dominant class per image)  
+
+---
+
+## Conversion Process
+
+1. **Source:** Semantic masks from `labelIds/CropAndWeed/`
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Both masks and bounding boxes have the category we need. I suppose it might be more easy to work with bounding boxes than with masks

File: docs/dataset_preparation.md [hunk 3]
```diff
@@ -1,45 +1,175 @@
...
+- Distribution: Crops (maize, sugar beet, pea) + weeds (bindweed, cockspur grass, wild carrot, etc.)  
+- Split: 80/20 (train/val) with PyTorch `random_split`  
+
+---
+
+## Setup Instructions
+
+### Install Requirements
+```bash
+pip install requests tqdm opencv-python torch torchvision
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Since we are working from inside the docker environment, we don't need these steps. If you need to install smth, just add this to `requirements.txt`

File: docs/dataset_preparation.md [hunk 4]
```diff
@@ -1,45 +1,175 @@
...
+```
+
+### Prepare Dataset
+```bash
+bash prepare_dataset.sh
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Once you are sure that it's working from within the docker env, please add a note that this command should be executed from the inside of the container

File: docs/dataset_preparation.md [hunk 5]
```diff
@@ -1,45 +1,175 @@
...
+
+Or manually:
+```bash
+python dataset_preparation/dataset_prep.py --data_root ./dataset
+python dataset_preparation/map_dataset.py --dataset_target CropsOrWeed9
+```
+
+---
+
+## Split Strategy ✅
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Let's not split it now - anyway we have zero idea how was the data split for training.

File: docs/dataset_preparation.md [hunk 6]
```diff
@@ -1,45 +1,175 @@
...
+- **Recommended:** 70/15/15 (train/val/test)  
+- **Why:** Separate test set → unbiased evaluation  
+- **Implementation:** PyTorch `random_split(seed=42)`  
+
+---
+
+## Challenges & Solutions
+
+### 1. Label Mapping  
+- **Problem:** Pixel mask → single label
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: already mentioned that it might be easier to use bboxes for that

File: docs/dataset_preparation.md [hunk 7]
```diff
@@ -1,45 +1,175 @@
...
+
+## Challenges & Solutions
+
+### 1. Label Mapping  
+- **Problem:** Pixel mask → single label  
+- **Solution:** Dominant class approach  
+
+### 2. Class Imbalance  
+- **Problem:** Few samples in some classes  
+- **Solution:** Filtered → 83 balanced classes
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Please specify how exactly did you filter, and what are the removed categories (and why removed)

File: docs/dataset_preparation.md [hunk 8]
```diff
@@ -1,45 +1,175 @@
...
+
+### 1. Label Mapping  
+- **Problem:** Pixel mask → single label  
+- **Solution:** Dominant class approach  
+
+### 2. Class Imbalance  
+- **Problem:** Few samples in some classes  
+- **Solution:** Filtered → 83 balanced classes  
+
+### 3. Data Loading Performance
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Let's remove this point - it's clear from the stage

File: docs/dataset_preparation.md [hunk 9]
```diff
@@ -1,45 +1,175 @@
...
+
+### 3. Data Loading Performance  
+- **Problem:** Large dataset slow to load  
+- **Solution:** Optimized transforms, batch sizes, dataloading  
+
+---
+
+## Evaluation Strategy
+
+- **Metrics:** Accuracy, Loss (TensorBoard)
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Why tensorboard? We don't know that

File: docs/dataset_preparation.md [hunk 10]
```diff
@@ -1,45 +1,175 @@
...
+- **Problem:** Large dataset slow to load  
+- **Solution:** Optimized transforms, batch sizes, dataloading  
+
+---
+
+## Evaluation Strategy
+
+- **Metrics:** Accuracy, Loss (TensorBoard)  
+- **Validation:** 20% held-out data  
+- **Models tested:** MobileNetV2, ResNet18, ShuffleNet
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: What shufflenet?

File: docs/dataset_preparation.md [hunk 11]
```diff
@@ -1,45 +1,175 @@
...
+
+---
+
+## Evaluation Strategy
+
+- **Metrics:** Accuracy, Loss (TensorBoard)  
+- **Validation:** 20% held-out data  
+- **Models tested:** MobileNetV2, ResNet18, ShuffleNet  
+
+**Future Improvements:**
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Let's not mention it in here

File: docs/dataset_preparation.md [hunk 12]
```diff
@@ -1,45 +1,175 @@
...
+**Future Improvements:**  
+- Add **test set (70/15/15)**  
+- Additional metrics: Precision, Recall, F1-score  
+- Confusion matrix per-class  
+
+---
+
+## Results ✅
+
+### Training
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: We don't train the model, please remove it

File: docs/dataset_preparation.md [hunk 13]
```diff
@@ -1,45 +1,175 @@
...
+- **Models tested:** MobileNetV2, ResNet18, ShuffleNet  
+
+**Future Improvements:**  
+- Add **test set (70/15/15)**  
+- Additional metrics: Precision, Recall, F1-score  
+- Confusion matrix per-class  
+
+---
+
+## Results ✅
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: I'd remove this points "Results"

File: docs/dataset_preparation.md [hunk 14]
```diff
@@ -1,45 +1,175 @@
...
+
+## Current Status
+- **Format Conversion:** Done (classification)  
+- **Split:** 80/20 (future: 70/15/15)  
+- **Training:** Models trained, saved in `models/`  
+- **Inference:** Working + documented  
+
+---
+
+## Summary
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: This section to remove as well

File: scripts/preprocess_data.sh [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+echo "=== Step 3: Done. Dataset ready at: $DATA_ROOT/classifier_data ==="
```
Comment (lyuzinmaxim) [inline scripts/preprocess_data.sh]: it's not a step; it's an outcome

File: scripts/preprocess_data.sh (hunk 1 above)
Comment (lyuzinmaxim) [reply scripts/preprocess_data.sh]: Are these 2 steps enough to run evaluation on the dataset?

**Comment by chani0343** [comment]
Hi ,

Thank you for all the feedback. I've updated the documentation based on your comments:

**Changes made:**
- ✅ Removed all references to CropsOrWeed9 mapping (not needed)
- ✅ Emphasized bounding boxes over semantic masks (easier to work with)
- ✅ Updated Docker environment instructions (execute inside container)
- ✅ Modified split strategy (use dataset as-is for evaluation)
- ✅ Focused on evaluation approach rather than training

**Key points addressed:**
- No mapping step needed - dataset is already properly labeled
- Bounding boxes are the recommended approach for label extraction
- Multi-label challenge documented with solution options
- Docker-first approach for setup instructions

The documentation now clearly focuses on dataset inspection and evaluation strategy as required by the ticket.

Ready for marge!

File: docs/dataset_preparation.md [hunk 16]
```diff
@@ -41,135 +41,104 @@ dataset/
 
 ---
 
-## Final Decision: Classification Structure ✅
+## Data Analysis Results ✅
 
-**Decision Made:** After analysis, we converted to a **classification structure**.  
+### Dataset Splits
+**Answer:** The original dataset does **not** provide predefined train/val/test splits. We need to create them ourselves.
```
Comment (r83575) [inline docs/dataset_preparation.md:47]: There is no question to answer here.
Please just explain it.

File: docs/dataset_preparation.md [hunk 17]
```diff
@@ -41,135 +41,104 @@ dataset/
...
+- Recommended by project lead
 
-**Statistics:**
-- Total images processed: Full dataset  
-- Active classes: **83** (see `src/class_mapping.py`)  
-- Distribution: Crops (maize, sugar beet, pea) + weeds (bindweed, cockspur grass, wild carrot, etc.)  
-- Split: 80/20 (train/val) with PyTorch `random_split`  
+### Approach
+- Extract labels from bounding box CSV files
+- Assign single label per image (dominant class by bbox area or count)
```
Comment (r83575) [inline docs/dataset_preparation.md:77]: The strategy is been changed, Please address it.

File: docs/dataset_preparation.md (hunk 16 above)
Comment (Rut-Vahab) [reply docs/dataset_preparation.md:47]: Great, we fixed it.

File: docs/dataset_preparation.md [hunk 18]
```diff
@@ -1,45 +1,137 @@
...
+## Current Status
+- **Dataset Downloaded:** ✅ 
+- **Structure Analyzed:** ✅
+- **Labels Identified:** ✅ (bounding boxes + semantic masks)
+- **Multi-label Issue:** ✅ (documented with solution options)
+- **Conversion Decision:** ✅ (use bounding boxes, keep original structure)
+
+---
+
+## Next Steps
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: I wouldn't add the information about the next steps to README, since for that we have our board.

File: docs/dataset_preparation.md [hunk 19]
```diff
@@ -1,45 +1,137 @@
...
+1. Implement bounding box label extraction
+2. Test evaluation pipeline with selected metrics
+3. Document filtering options for different dataset sizes
+
+---
+
+## Summary
```
Comment (lyuzinmaxim) [inline docs/dataset_preparation.md]: Also I wouldn't add the "summary" part since it's selfexplanatory

File: dataset_preparation/dataset_prep.py [hunk 20]
```diff
@@ -21,6 +21,9 @@ def parse_arguments():
 
 
 def setup(data_root: str):
+    import urllib3
```
Comment (lyuzinmaxim) [inline dataset_preparation/dataset_prep.py]: What is the purpose of it?

Comment (chani0343) [reply dataset_preparation/dataset_prep.py]: I needed to download the data and I had an SSL problem and I asked GPT and he told me to add this line and that solved my problem
We want to move on to the next task

Comment (lyuzinmaxim) [reply dataset_preparation/dataset_prep.py]: since we are running this function each time we're running the script, let's move the import to the top of the file

Comment (Rut-Vahab) [reply dataset_preparation/dataset_prep.py]: Ok thanks we'll move it..
Can we merge this now?
